chain/chain/src/chain.rs

```diff
@@ -1038,6 +1041,94 @@ impl Chain {
+            let mut all_parents_known = true;
+            let mut prev_hash = orphan.block.header.prev_hash;
+            for _ in 0..NUM_ORPHAN_PARENTS_CHECK {
+                if prev_hash == CryptoHash::default() {
+                    // TODO This is very suspicious.
+                    // Technically, it's correct and also helps in testing.
+                    // In reality it show never happen except maybe first seconds
+                    // of creating Universe.
+                    break;
+                }
+                // 1. Look into store
+                match self.store.get_block(&prev_hash) {
+                    Ok(block) => {
+                        accepted_parent = true;
+                        accepted_parent_hash = prev_hash;
+                        prev_hash = block.header.prev_hash;
+                        continue;
+                    }
+                    _ => {}
+                }
+                // 2. Look into missing chunks
+                if self.blocks_with_missing_chunks.orphans.contains_key(&prev_hash) {
+                    let prev_orphan =
+                        self.blocks_with_missing_chunks.orphans.get(&prev_hash).unwrap();
+                    prev_hash = prev_orphan.block.header.prev_hash;
+                    continue;
+                }
+                // 3. Look into orphans
+                if self.orphans.orphans.contains_key(&prev_hash) {
+                    let prev_orphan = self.orphans.orphans.get(&prev_hash).unwrap();
+                    prev_hash = prev_orphan.block.header.prev_hash;
+                    continue;
+                }
+                // Parent is not found
+                all_parents_known = false;
+                break;
+            }
+            // The following conditions should be fulfilled to request chunks:
+            // 1. At least one of the parents is accepted (is in store)
+            // 2. All of the parents_to_check should be known (in store, missing chunks or orphans)
+            if accepted_parent && all_parents_known {
+                orphan_hashes.push((accepted_parent_hash, orphan_hash));
+            }
+        }
+        if self.orphans.orphans.len() > 0 {
+            debug!(target: "chain", "Orphans total: {:?}, orphans to get chunks: {:?}", self.orphans.orphans.len(), orphan_hashes.len());
+        }
+
+        let mut chain_update = ChainUpdate::new(
+            &mut self.store,
+            self.runtime_adapter.clone(),
+            &self.orphans,
+            &self.blocks_with_missing_chunks,
+            self.transaction_validity_period,
+            self.epoch_length,
+            &self.block_economics_config,
+        );
+        for (parent_hash, orphan_hash) in orphan_hashes.into_iter() {
+            let orphan = self.orphans.orphans.get(&orphan_hash).unwrap();
+            // TODO check is it okay to use parent_hash here?
+            match chain_update.ping_missing_chunks(me, parent_hash, &orphan.block) {
```

Comment:
If it OK to use `parent_hash` for as long as the block is in the same epoch as the `parent_hash`.
So makes sense to check it here, and not proceed with the orphan if it's in the next epoch.